src/ql/arch/cc/pass/gen/vq1asm/detail/codegen.cc

```diff
@@ -544,250 +724,432 @@ void Codegen::customGate(
+    cs.emit(as_label(to_end(label)));    // label for loop end or 'break'
 }
 
-void Codegen::doWhileEnd(const Str &label, UInt op0, const Str &opName, UInt op1) {
-    comment(QL_SS2S("# DO_WHILE_END(R" << op0 << " " << opName << " R" << op1 << ")"));
-    emit("", "jmp", QL_SS2S("@" << label), "# FIXME: we don't support conditions, just an endless loop'");        // FIXME: just endless loop
-    QL_WOUT("CC backend ignores condition of do while loop");
-#if OPT_PRAGMA
-    emit((label+"_end:"), "", "", "# ");    // label for 'break'
-    pragmaLoopLabel.pop_back();
-#endif
+// NB: also used for 'while' loops
+void Codegen::for_start(utils::Maybe<ir::SetInstruction> &initialize, const ir::ExpressionRef &condition, const Str &label) {
+    comment(
+        "# LOOP_START: "
+        + (!initialize.empty() ? "initialize = '"+ir::describe(initialize)+"', " : "")
+        + "condition = '" + ir::describe(condition) + "'"
+    );
+
+    // for loop: initialize
+    if (!initialize.empty()) {
+        handle_set_instruction(*initialize, "for.initialize");
+        cs.emit("", "nop");    // register dependency between initialize and handle_expression (if those use the same register, which is likely)
+    }
+
+    cs.emit(as_label(to_start(label)));    // label for looping or 'continue'
+    handle_expression(condition, to_end(label), "for/while.condition");
+}
+
+
+void Codegen::for_end(utils::Maybe<ir::SetInstruction> &update, const Str &label) {
+    comment(
+        "# LOOP_END: "
+        + (!update.empty() ? " update = '"+ir::describe(update)+"'" : "")
+    );
+    if (!update.empty()) {
+        handle_set_instruction(*update, "for.update");
+    }
+    cs.emit("", "jmp", as_target(to_start(label)), "# loop");
+    cs.emit(as_label(to_end(label)));    // label for loop end or 'break'
+}
+
+
+void Codegen::do_break(const Str &label) {
+    cs.emit("", "jmp", as_target(to_end(label)), "# break");
+}
+
+
+void Codegen::do_continue(const Str &label) {
+    cs.emit("", "jmp", as_target(to_start(label)), "# continue");
 }
 
+
 void Codegen::comment(const Str &c) {
-    if (options->verbose) emit(c);
+    if (options->verbose) {
+        cs.emit(Str(2*depth, ' ') + c);  // indent by depth
+    }
 }
 
 /************************************************************************\
-|
-| private functions
-|
+| new IR expressions
 \************************************************************************/
 
-/************************************************************************\
-| Some helpers to ease nice assembly formatting
-\************************************************************************/
+// FIXME: recursion?
+// FIXME: or pass SetInstruction or Expression depending on use
+// FIXME: adopt structure of cQASM's cqasm-v1-functions-gen.cpp register_into used for constant propagation
 
-// FIXME: assure space between fields!
-// FIXME: make comment output depend on verboseCode
+// FIXME: see expression_mapper.cc for inspiration
```

Comment:
Depending on the target's capabilities, there may be several walks over the expression here, even done by passes before this one, assuming that this codegen essentially only wants to do emits.
Examples:
when there is a mismatch between an expression node and a target function, then that expression node should be transformed to match the primitive target function first
allocation of intermediate registers, when needed, in general: register allocation
necessary conversion functions may have to be added
reading from a location may need address arithmetic to be done (which is also code)

Reply:
Not sure what you mean. The current implementation does not try to generate code for every conceivable expression, but it does try to either work correctly, or give an appropriate error message. For one thing this seems to cover most of our needs, on the other hand doing things like a 'real' compiler is beyond the current scope of work.